037-pluggable-pod-security-profiles.md

```diff
@@ -0,0 +1,196 @@
+
+Apart from the Kaniko builder used for the Kafka Connect build, all our components are able to run under the _restricted_ profile.
+But the operator doesn't configure the appropriate security context and the admission plugin will reject them.
+Users have to _manually_ add the matching security context configuration into the `.template` sections of the custom resources and into the operator deployments to allow the pods to be created under the _restricted_ profile.
+To match the _restricted_ profile, users would need to add the following security context:
+
+```yaml
+          securityContext:
+            allowPrivilegeEscalation: false
+            capabilities:
+              # The restricted allows the use of the NET_BIND_SERVICE capability.
+              # But Strimzi does not require it for anything, so we do not include it here.
+              drop:
+                - ALL
+            runAsNonRoot: true
+            seccompProfile:
+              type: RuntimeDefault
+```
+
+While users can configure the security context, it is not user-friendly.
+The Pod security context does not cover all the different options required by the _restricted_ profile.
+So the security context needs to be basically configured for every container (including init containers, sidecars etc.).
+So just for a single Kafka cluster, it might need to be configured on 9 different places of the Kafka CR (ZooKeeper, Kafka broker, Kafka init-container, Topic Operator, User Operator, TLS sidecar, Kafka Exporter, Cruise Control, another TLS sidecar).
+
+## Proposal
+
+This proposal suggests to add a pluggable mechanism for users to provide their own plugins to configure the Pod and container security contexts of all Strimzi components.
+Out of the box, Strimzi will provide implementation supporting two profiles:
+* The _baseline_ provider which will mirror the current behavior and can be used when running under the _baseline_ Kubernetes profile
+* The _restricted_ provider which will allow to run under the _restricted_ Kubernetes profile
+
+The _baseline_ profile will be used by default.
+So users will not see any change by default.
+But if they decide to, they will be abe to switch to the _restricted_ profile.
+
+In addition, any user who will need some special configuration, will be able to provide a custom provider plugin.
+This will allow them to use their own configurations without specifying them in all custom resources in the `.template` sections.
+This might be useful for users who use some custom security tooling and might have their own requirement for the security configuration.
+But regular users will not need to write any custom plugins.
+
+The puggable mechanism should also make it easy for the Strimzi itself to add any additional implementations in the future.
+
+### Plugin implementation
+
+#### Interface
+
+A new interface `PodSecurityProvider` will be added to the `api` module.
```

Comment:
Having a single interface might cause difficulties for implementors when we want to add or remove pods. Using `default` methods helps with adding. But removing a method would break source compatibility when implementers had use `@Override`. Maybe having a separate interface for each kind of pod would make for a better compatibility story.

Reply:
TBH, I don't like this approach. The interface has many methods, but they in general do the same. While we want to provide the granularity _just in case it is needed_, I expect that in most cases they will just delegate to a private method. This is also what the implementations of the baseline and restricted profiles which should be part of Strimzi do.

Having a separate interface for each operand would mean a lot of interfaces and lot of implementation classes. They will largely do all the same but just with higher maintenance effort. Also, the _injection_ of the providers into the model classes in the `cluster-operator` model will need to be done multiple times instead of just once.